File: SDK_Examples/C#/Examples/ReadCard.cs
using System;
using pt.portugal.eid;

namespace Examples
{
    class ReadCard
    {
        //Main attributes needed for SDK functionalities
        PTEID_ReaderSet readerSet = null;
        PTEID_ReaderContext readerContext = null;
        PTEID_EIDCard eidCard = null;
        PTEID_EId eid = null;
        PTEID_CardType cardType = PTEID_CardType.PTEID_CARDTYPE_UNKNOWN;
        PTEID_CardContactInterface contactInterface = PTEID_CardContactInterface.PTEID_CARD_CONTACTEMPTY;


        /*
         * Initializes the SDK and sets main variables
         */
        public void Initiate()
        {
            //Must always be called in the beginning of the program
            PTEID_ReaderSet.initSDK();

            //Sets test mode to true so that CC2 can be tested
            //PTEID_Config.SetTestMode(true);

            //Gets the set of connected readers, if there is any inserted
            readerSet = PTEID_ReaderSet.instance();

            //Gets the first reader
            //When multiple readers are connected, you should iterate through the various indexes with the methods getReaderName and getReaderByName
            readerContext = readerSet.getReader();

            //Gets the Card Contact Interface and type

            if (readerContext.isCardPresent())
            {
                contactInterface = readerContext.getCardContactInterface();
                cardType = readerContext.getCardType();
                Console.WriteLine("Contact Interface:" + (contactInterface == PTEID_CardContactInterface.PTEID_CARD_CONTACTLESS ? "CONTACTLESS" : "CONTACT"));
            }

            //Gets the card instance
            eidCard = readerContext.getEIDCard();

            //If the contactInterface is contactless and the card supports contactless then authenticate with PACE
            if (contactInterface == PTEID_CardContactInterface.PTEID_CARD_CONTACTLESS && cardType ==  PTEID_CardType.PTEID_CARDTYPE_IAS5){
                Console.WriteLine("Insert the CAN for this EIDCard: ");
                string can_str = Console.ReadLine();
                uint can_size = (uint) can_str.Length;
                eidCard.initPaceAuthentication(can_str, can_size,  PTEID_CardPaceSecretType.PTEID_CARD_SECRET_CAN);
            }
            eid = eidCard.getID();
        }

        /*
         * Releases the SDK (must always be done at the end of the program)
         */
        public void Release()
        {
            try
            {
                PTEID_ReaderSet.releaseSDK();
            }
            catch (PTEID_Exception ex)
            {
                Console.WriteLine(ex.Message);
            }
        }

        /*
         * Prints main info present in the card (that doesn't need address pin)
         */
        public void ShowCardInfo()
        {
            Console.WriteLine("\n\nReading card details:");
            Console.WriteLine("Name:                       " + eid.getGivenName() + " " + eid.getSurname());
            Console.WriteLine("Card Type:                  " + eid.getDocumentType());
            Console.WriteLine("Card Version:               " + eid.getDocumentVersion());
            Console.WriteLine("Card Number:                " + eid.getDocumentNumber());
            Console.WriteLine("Local of Request:           " + eid.getLocalofRequest());
            Console.WriteLine("Issuing Entity:             " + eid.getIssuingEntity());
            Console.WriteLine("Issuing Date:               " + eid.getValidityBeginDate());
            Console.WriteLine("Validity End Date:          " + eid.getValidityEndDate());
            Console.WriteLine("PAN Number:                 " + eid.getDocumentPAN());
            Console.WriteLine("Civilian ID :               " + eid.getCivilianIdNumber());
            Console.WriteLine("Tax ID:                     " + eid.getTaxNo());
            Console.WriteLine("Social Security ID:         " + eid.getSocialSecurityNumber());
            Console.WriteLine("Health ID:                  " + eid.getHealthNumber());
            Console.WriteLine("Parents:                    " + eid.getParents());
            Console.WriteLine("Father:                     " + eid.getGivenNameFather() + " " + eid.getSurnameFather());
            Console.WriteLine("Mother:                     " + eid.getGivenNameMother() + " " + eid.getSurnameMother());
            Console.WriteLine("Indications:                " + eid.getAccidentalIndications());
            Console.WriteLine("Nationality:                " + eid.getNationality());
            Console.WriteLine("Country:                    " + eid.getCountry());
            Console.WriteLine("Date of birth:              " + eid.getDateOfBirth());
            Console.WriteLine("Height:                     " + eid.getHeight());
            Console.WriteLine("Gender:                     " + eid.getGender());
            Console.WriteLine("MRZ (Machine Readable Zone): " + eid.getMRZ1());
            Console.WriteLine("                             " + eid.getMRZ2());
            Console.WriteLine("                             " + eid.getMRZ3());
            Console.WriteLine((cardType == PTEID_CardType.PTEID_CARDTYPE_IAS5 ? "CC2" : "CC1"));
        }

        public void start()
        {
            try
            {
                Initiate();
                ShowCardInfo();
            }
            catch (PTEID_ExNoReader)
            {
                Console.WriteLine("No reader found.");
            }
            catch (PTEID_ExNoCardPresent)
            {
                Console.WriteLine("No card inserted.");
            }
            catch (PTEID_Exception ex)
            {
                Console.WriteLine(ex.GetMessage());
            }
            finally {
                Release();
                Console.ReadLine();
            }
        }   
    
        static void Main(string[] args)
        {
            new ReadCard().start();
        }
    }
}

File: SDK_Examples/C++/ReadCard.cpp
#include "eidlib.h"
#include "eidlibException.h"
#include <iostream>

using namespace eIDMW;

int main(int argc, char **argv) {

    try 
    {
        //Initializes SDK (must always be called in the beginning of the program)
        PTEID_InitSDK();

        //Gets the set of readers connected to the system
        PTEID_ReaderSet& readerSet = PTEID_ReaderSet::instance();

        unsigned long reader_count = ReaderSet.readerCount();
        const char * const * reader_list = ReaderSet.readerList();
        std::cout << "Connected Readers: " << std::endl;

        for (int i=0; i < reader_count; i++) {
            std::cout << i << ": " << reader_list[i] << std::endl;
        }

        //Gets a reader connected to the system (useful if you only have one)
        //Alternatively you can iterate through the readers using getReaderByNum(int index) instead of getReader()
        PTEID_ReaderContext& reader = readerSet.getReader();

        //Gets the EIDCard and EId objects (with the cards information)
        PTEID_EIDCard& eidCard = reader.getEIDCard();
        PTEID_EId& eid = eidCard.getID();

        std::cout << "Name:                       " << eid.getGivenName() << " " << eid.getSurname() << std::endl;
        std::cout << "Card Type:                  " << eid.getDocumentType() << std::endl;
        std::cout << "Card Version:               " << eid.getDocumentVersion() << std::endl;
        std::cout << "Card Number:                " << eid.getDocumentNumber() << std::endl;
        std::cout << "Local of Request:           " << eid.getLocalofRequest() << std::endl;   
        std::cout << "Issuing Entity:             " << eid.getIssuingEntity() << std::endl;
        std::cout << "Issuing Date:               " << eid.getValidityBeginDate() << std::endl;
        std::cout << "Validity End Date:          " << eid.getValidityEndDate() << std::endl;
        std::cout << "PAN Number:                 " << eid.getDocumentPAN() << std::endl;
        std::cout << "Civilian ID :               " << eid.getCivilianIdNumber() << std::endl;
        std::cout << "Tax ID:                     " << eid.getTaxNo() << std::endl;
        std::cout << "Social Security ID:         " << eid.getSocialSecurityNumber() << std::endl;
        std::cout << "Health ID:                  " << eid.getHealthNumber() << std::endl;
        std::cout << "Parents:                    " << eid.getParents() << std::endl;
        std::cout << "Father:                     " << eid.getGivenNameFather() << " " << eid.getSurnameFather() << std::endl;
        std::cout << "Mother:                     " << eid.getGivenNameMother() << " " << eid.getSurnameMother() << std::endl;
        std::cout << "Indications:                " << eid.getAccidentalIndications() << std::endl;
        std::cout << "Nationality:                " << eid.getNationality() << std::endl;
        std::cout << "Country:                    " << eid.getCountry() << std::endl;
        std::cout << "Date of birth:              " << eid.getDateOfBirth() << std::endl;
        std::cout << "Height:                     " << eid.getHeight() << std::endl;
        std::cout << "Gender:                     " << eid.getGender() << std::endl;
        std::cout << "MRZ (Machine Readable Zone): " << eid.getMRZ1() << std::endl;
        std::cout << "                             " << eid.getMRZ2() << std::endl;
        std::cout << "                             " << eid.getMRZ3() << std::endl;   
    }
    catch (PTEID_ExNoReader &e) 
    {
        std::cerr << "No readers found!" << std::endl;
    }
    catch (PTEID_ExNoCardPresent &e) 
    {
        std::cerr << "No card found in the reader!" << std::endl;
    }
    catch (PTEID_Exception &e) 
    {
        std::cerr << "Caught exception in some SDK method. Error: " << e.GetMessage() << std::endl;
    }

    //Releases SDK (must always be called at the end of the program)
    PTEID_ReleaseSDK();
    return 0;
}
